src/asset_manager/mod.rs

```diff
@@ -1,4 +1,287 @@
 //! Asset manager used to load assets (like `Mesh`es and `Texture`s).
+//!
+//! For how to implement an asset yourself, see the `Asset` trait.
+//!
+//! If you just want to load it, look at `AssetLoader` / `AssetManager`.
 
-mod asset_manager;
-pub use self::asset_manager::*;
+pub mod formats;
+
+mod asset;
+mod common;
+mod io;
+
+pub use self::asset::{Asset, AssetFormat, AssetStore, AssetStoreError};
+pub use self::common::{DefaultStore, DirectoryStore, ZipStore};
+pub use self::io::{Import, Error as ImportError};
+
+use std::fmt::{Debug, Display, Error as FormatError, Formatter};
+use std::marker::Send;
+
+use futures::{Async, Future};
+use futures_cpupool::{CpuPool, CpuFuture};
+
+use engine::Context;
+
+/// A parallel asset loader.
+/// It contains a `CpuPool` and
+/// creates `Future`s from that.
+///
+/// Only the `Asset::Data` will be
+/// loaded in a separate thread, the
+/// asset itself will be created on the main
+/// thread, because often it requires thread-unsafe
+/// actions (accessing OpenGL to create some buffer).
+///
+/// The asset loader needs three things in order to
+/// load an asset:
+///
+/// * An asset store (`AssetStore`): Responsible for providing bytes for a name and a format
+/// * The asset name: A simple `&str`
+/// * The asset format (`AssetFormat`): Has to implement `Import` (create the `Asset::Data` structure) and provide the typical file extension
+///
+/// # Examples
+///
+/// ```ignore
+/// use amethyst::asset_manager::AssetLoader;
+///
+/// let loader = AssetLoader::new();
+/// let asset_future = loader.load(my_store, "my_asset", MyFormat);
+/// ```
+pub struct AssetLoader {
+    cpupool: CpuPool,
+}
+
+/// An asset future which means
+/// "This will be available in the future".
+///
+/// #### Why isn't it available immediately?
+///
+/// The reason is that assets are loaded in
+/// separate threads, without blocking the
+/// main thread. Once they're finished,
+/// you can access it.
+///
+/// As soon as you need the asset, you call
+/// `finish` on this future which will return
```

Comment:
In order to have asset loading that doesn't block the main thread [Future::poll](https://docs.rs/futures/0.1.10/futures/future/trait.Future.html#tymethod.poll) should be used instead of [Future::wait](https://docs.rs/futures/0.1.10/futures/future/trait.Future.html#method.wait) (it is used inside `AssetFuture::finish`).

Probably the best way to achieve this would be to add a `AssetLoader` to `Engine` struct. Then `AssetFuture::wait` can be called repeatedly from `State::update` (as described above in my `03_renderable` example comment). Then this doc comment would have to be updated to reflect the proper usage.